webextension/background.js

```diff
@@ -6,14 +6,136 @@
 
 /* globals browser */
 
-const gURLThankYou = "https://mozilla.github.io/webcompat-blipz-experiment/thanks.html";
+let gCurrentlyPromptingTabId;
+
+const Config = (function() {
+  class Config {
+    constructor() {
+      this._neverShowAgain = false;
+      this._skipPrivateBrowsingTabs = true;
+      this._lastPromptTime = 0;
+      this._domainsToCheck = {
+        "accounts.google.com": 0,
+        "amazon.com": 0,
+        "amazon.in": 0,
+        "bing.com": 0,
+        "docs.google.com": 0,
+        "drive.google.com": 0,
+        "facebook.com": 0,
+        "flipkart.com": 0,
+        "github.com": 0,
+        "google.co.in": 0,
+        "google.com": 0,
+        "inbox.google.com": 0,
+        "instagram.com": 0,
+        "linkedin.com": 0,
+        "mail.google.com": 0,
+        "netflix.com": 0,
+        "pandora.com": 0,
+        "play.google.com": 0,
+        "reddit.com": 0,
+        "soundcloud.com": 0,
+        "theverge.com": 0,
+        "twitch.tv": 0,
+        "twitter.com": 0,
+        "web.whatsapp.com": 0,
+        "youtube.com": 0,
+      };
+    }
 
-const gMinimumFrequencyBeforeRePrompting = 1000 * 20; // 20 seconds (for testing)
-const gSkipPrivateBrowsingTabs = true;
+    load() {
+      // TBD
+      return Promise.resolve();
+    }
 
-const gDomainCheckTimestamps = {};
+    save(options) {
+      // TBD
+      console.info("Saving options", options);
+    }
 
-let gNeverShowAgain = false;
+    onUserPrompted(domain) {
+      const now = Date.now();
+      this._lastPromptTime = now;
+      this._domainsToCheck[domain] = now;
+      this.save({
+        lastPromptTime: now,
+        domainsToMatch: this._domainsToCheck
+      });
+    }
+
+    findDomainMatch(domain) {
+      for (const candidate of Object.keys(this._domainsToCheck)) {
+        if (domain.endsWith(candidate)) {
+          return candidate;
+        }
+      }
+      return undefined;
+    }
+
+    shouldPromptUserNow(domain) {
+      // Prompt at most once a day, at a minimum once every three days,
+      // for a maximum of five prompts per user and one prompt per domain.
+      if (this._lastPromptTime) {
+        const now = Date.now();
+        const oneDay = 1000 * 60 * 60 * 24;
+        const nextValidCheckTime = this._lastPromptTime + oneDay;
+        const nextNecessaryCheckTime = this._lastPromptTime - (oneDay * 3);
+        if (now < nextValidCheckTime &&
+            (now > nextNecessaryCheckTime || Math.random() > 0.5)) {
```

Comment:
So we're saying, if it's too early to prompt, OR if it's OK to prompt and you roll the correct dice, we prompt you. I think that sounds good.

Reply:
One question I have, is at what point do we ask if we should prompt the user? How long after page load? What would the ideal value be?

Reply:
That is a very good question, and since we don't have an answer I'm just prompting right away. I think this deserves its own issue in the tracker.

We could hypothetically detect specific known-to-be-iffy HTML, CSS, JS, console messages, etc. For instance, if window.event is being accessed when it doesn't exist.

Or we could even wait for some signal that the user is "upset" (hopefully there is some research on how to detect that).

For now I'm guessing it might be best to just wait until the page has been loaded and the user has interacted with it, but has stopped for at least a few seconds (or perhaps if they're about to leave the page, in onunload or something).